_posts/2020-06-19-dask-usage-at-blue-yonder.markdown

```diff
@@ -0,0 +1,189 @@
+
+### Dataset maintenance
+
+We use Dask/Distributed for performing maintenance operations like garbage collection on datasets.
+Garbage collection is necessary since we use a lazy approach when updating datasets and only delete or update
+references to blobs in the dataset metadata, deferring the deletion of no longer used blobs.
+In this case, Dask is used to parallelize the delete requests to the storage provider (Azure blob service).
+In a similar fashion, we use Dask to parallelize copy operations for backing up datasets.
+
+### Repartitioning a dataset
+
+Our algorithms rely on source datasets that are partitioned in a particular way. It is not always
+the case that we have the data available in a suitable partitioning, for instance, if the data results
+from a computation that used a different partitioning. In this case, we have to repartition the data.
+This works by reading the dataset as a [Dask dataframe](https://docs.dask.org/en/latest/dataframe.html) 
+repartitioning the dataframe using network shuffle, and writing it out again to a dataset.
+
+## Ad hoc data analysis
+
+Finally, our data scientists also use Dask for out-of-core analysis of data using Jupyter notebooks.
+This allows us to take advantage of all the nice features of Jupyter even for data that is too large to
+fit into an individual compute node's memory.
+
+## Dask cluster setup at Blue Yonder
+
+We run a somewhat unique setup of Dask clusters that is driven by the specific requirements
+of our domain (environment isolation for SaaS applications and data isolation between customers). 
+We have a large number of individual clusters
+that are homogeneous in size with many of the clusters dynamically scaled. The peculiarities of this
+setup have in some cases triggered edge cases and uncovered bugs, which lead us to submit a number
+of upstream pull requests.
+
+### Cluster separation
+
+To provide isolation between customer environments we run separate Dask clusters per customer.
+For the same reason, we also run separate clusters for production, staging, and development environments. 
+
+But it does not stop there. 
+The service we provide to our customers is comprised of several products that build upon each other and 
+and maintained by different teams. We typically perform daily batch runs with these products running sequentially
+in separated environments.
+For performance reasons, we install the Python packages holding the code needed for the computations on each worker. 
+We do not want to synchronize the dependencies and release cycles of our different products, which
+means we have to run a separate Dask cluster for each of the steps in the batch run.
+This results in us operating more than
+ten Dask clusters per customer and environment, with most of the time, only one of the clusters being active
+and computing something. While this leads to overhead in terms of administration and hardware resouces,
```

Comment:
This might be a nice opportunity to remind readers of the dynamic up/downscaling